### PR TITLE
Support custom Taskcluster routes

### PR DESCRIPTION
**Pull request by La0**
The idea here is to index Taskcluster tasks triggered by pulse listener, as soon as they are created, and using extra data from `triggerHook` payload.

Static analysis would use this feature to index tasks with their analysis ID (mozreview ID or phabricator DIFF ID)

### CODE REVIEW

File: src/shipit_static_analysis/default.nix [hunk 1]
```diff
@@ -49,6 +52,26 @@ let
...
+
+        taskRoutes = {
+          "$flatten" = [
+            [
+              # Latest route
+              ("index.project.releng.services.project." + branch + ".shipit_static_analysis.latest")
+
+            ]
+            {
+              "$if" = "firedBy == 'triggerHook'";
```
Comment (marco-c) [inline src/shipit_static_analysis/default.nix]: Oh right, so we could apply the same solution to code coverage, by using firedBy == 'schedule' instead. Could you make the changes to code coverage too?

File: src/shipit_pulse_listener/shipit_pulse_listener/listener.py [hunk 2]
```diff
@@ -133,14 +134,18 @@ def parse(self, body):
             return
 
         # Extract commits
-        commits = [
-            '{rev}:{review_request_id}:{diffset_revision}'.format(**c)
+        envs = [
```
Comment (marco-c) [inline src/shipit_pulse_listener/shipit_pulse_listener/listener.py]: This is going to increase the workload for mozreview, but I guess we don't really care. Hopefully when we have shallow clones for mozilla-central the impact will be smaller, and mozreview is going to be retired at some point anyways.

File: src/shipit_pulse_listener/shipit_pulse_listener/hook.py [hunk 3]
```diff
@@ -93,6 +93,10 @@ def build_consumer(self, pulse_user, pulse_password):
         env = self.parse(body)
         if env is None:
             logger.warn('Skipping message, no task created', hook=self.hook_id)
+        elif isinstance(env, list):
```
Comment (marco-c) [inline src/shipit_pulse_listener/shipit_pulse_listener/hook.py]: *Nit*: we could make env a list by default (and call it envs), to reduce the amount of code here.

Comment (La0) [reply src/shipit_pulse_listener/shipit_pulse_listener/hook.py]: Done !

File: src/shipit_static_analysis/shipit_static_analysis/workflow.py [hunk 4]
```diff
@@ -286,3 +308,35 @@ def build_improvement_patch(self, revision, issues):
...
         logger.info('Diff available online', url=revision.diff_url)
+
+    def index(self, revision, **kwargs):
+        '''
+        Index current task on Taskcluster index
+        '''
+        assert isinstance(revision, Revision)
+
+        if not self.on_taskcluster:
+            logger.info('Skipping taskcluster indexation', rev=str(revision), **kwargs)
```
Comment (marco-c) [inline src/shipit_static_analysis/shipit_static_analysis/workflow.py]: Nit: it should be `indexing` instead of `indexation`

Comment (marco-c) [reply src/shipit_static_analysis/shipit_static_analysis/workflow.py]: Here and in other places

Comment (La0) [reply src/shipit_static_analysis/shipit_static_analysis/workflow.py]: :fr: FTW

File: src/shipit_static_analysis/shipit_static_analysis/workflow.py [hunk 5]
```diff
@@ -202,10 +218,14 @@ def run(self, revision):
...
         if not issues:
             logger.info('No issues, stopping there.')
+            self.index(revision, state='done', issues=0)
             return
 
         # Report issues publication stats
-        stats.api.increment('analysis.issues.publishable', len([i for i in issues if i.is_publishable()]))
+        nb_issues = len(issues)
+        nb_publishable = len([i for i in issues if i.is_publishable()])
+        self.index(revision, state='analyzed', issues=nb_issues, issues_publishable=nb_publishable)
```
Comment (marco-c) [inline src/shipit_static_analysis/shipit_static_analysis/workflow.py]: Maybe we are using the index service not exactly in its intended use case as we're storing data of the tasks too and we're using it to report progress.
In the future, we might want to look into the [taskcluster-treeherder](https://docs.taskcluster.net/docs/reference/integrations/taskcluster-treeherder) service.

**Pull request merged**